### PR TITLE
test(e2e): add end to end test setup with webdriver-io

### LINKED ISSUE #35: End-to-end testing pipeline
Resources:

- https://code.visualstudio.com/api/extension-guides/testing
- https://code.visualstudio.com/api/working-with-extensions/testing-extension
- https://github.com/microsoft/vscode-extension-samples/tree/main/helloworld-test-sample

Screenshot 2023-10-25 at 12 00 49 PM

### PR DESCRIPTION
**Pull request by QZera**
Closes #35

### CODE REVIEW

**Review by maninak — DISMISSED** [review]

File: test/specs/onboarding.e2e.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    })
+  })
+
+  describe('after the repo has been initialized as a radicle repo', () => {
+    let cliCommandsSection: ViewSection
+
+    before(async () => {
+      await $`rad init --private --default-branch main --name "A_test_blog" --description "Some repo" --no-confirm --verbose`
+      await openRadicleSidebarTab(workbench)
+      const sidebarView = workbench.getSideBar().getContent()
```
Comment (maninak) [inline test/specs/onboarding.e2e.ts:47]: it's an Activity Bar. It may, or may not be positioned on the side, depending on user preference.

File: package.json [hunk 2]
```diff
@@ -49,7 +49,7 @@
     "test:typings:extension": "tsc --noEmit",
     "test:typings:webviews": "cd ./src/webviews && npm run type-check",
     "test:unit": "echo \"Unit testing not yet implemented.\"",
-    "test:e2e": "echo \"End-to-end testing not yet implemented.\""
+    "test:e2e": "wdio run ./test/wdio.conf.ts"
```
Comment (maninak) [inline package.json:52]: Should we adjust L47 above to *not* contain `&& npm run test:e2e`? Or have it to only run if env var `CI` is set?

Ugh.. not being able to run those tests locally is such a bummer. What if we (for some scenarios?) would install radicle in specific destination e.g. under [`os.tempDir()`](https://nodejs.org/api/os.html#os_os_tmpdir) with

```sh
curl -sSf https://radicle.xyz/install | sh -s -- --prefix=my/temp/dir --no-modify-path
```

According to the installation script:

> `--prefix=PATH       Radicle install prefix (default: ~/.radicle)`

File: .github/workflows/end-to-end-tests.yaml
Comment (maninak) [inline .github/workflows/end-to-end-tests.yaml:1]: I think to align all jobs names and filenames we should

- name this job `test-e2e` and its file `e2e-testing.yaml`
- rename ci.yaml as `code-linting.yaml` and its "audit" job as `lint-code` and remove the inert "name: Test Unit" step
- rename file pr-linting.yaml as `lint-pr.yaml` and change the `name` definition to be `CI`
- rename file commit-linting.yaml as `lint-commits.yaml` and change the `name` definition to be `CI`

I hope I didn't mess something up here :P

The above should result in more homogenously named files/jobs/tasks and a cleaner listing on GitHub UI, where everything under "CI" is coming from within our `.github/workflows` dir.

image

**Review by maninak — APPROVED** [review]
So great to have these first steps in the repo, thank you!

Most of the things I found are regarding convensions. For example 
I'm not sure if the `test` dir, given that it contains a fine-tuned tsconfig.json etc should be better named `test-e2e` or perhaps it should be a subfolder like `test/e2e/` or just left as is.

Especially in the context of eventual unit tests against extension util logic, etc.

**Review by maninak — CHANGES_REQUESTED** [review]
Sry meant to request changes with the previous one.

File: .github/workflows/end-to-end-tests.yaml
Comment (maninak) [reply .github/workflows/end-to-end-tests.yaml:1]: I already spotted a mistake in my specs here :man_facepalming:. Oh well. Resolved.

File: test/wdio.conf.ts [hunk 3]
```diff
@@ -0,0 +1,321 @@
+import path from 'node:path'
+import { e2eTestDirPath } from './constants/config'
+
```
[the remark this answers is not in the document]
Comment (QZera) [reply test/wdio.conf.ts:3]: Good point!

Opted to throw an error on this file to keep the package.json script clean. I think throwing an error and bailing makes more sense here to be safe.

File: test/wdio.conf.ts [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+  // - @wdio/cli, @wdio/config, @wdio/utils
+  // Level of logging verbosity: trace | debug | info | warn | error | silent
+  // logLevels: {
+  //     webdriver: 'info',
+  //     '@wdio/appium-service': 'info'
+  // },
+  //
+  // If you only want to run your tests until a specific amount of tests have failed use
+  // bail (default is 0 - don't bail, run all tests).
+  bail: 0,
```
[the remark this answers is not in the document]
Comment (QZera) [reply test/wdio.conf.ts]: ✅

File: test/wdio.conf.ts [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+   * @param {object} params information about the assertion to be executed
+   */
+  // beforeAssertion: function(params) {
+  // }
+  /**
+   * Hook that gets executed after a WebdriverIO assertion happened.
+   * @param {object} params information about the assertion that was executed, including its results
+   */
+  // afterAssertion: function(params) {
+  // }
```
[the remark this answers is not in the document]
Comment (QZera) [reply test/wdio.conf.ts]: ✅

File: .github/workflows/end-to-end-tests.yaml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        rad node start &
+        echo "allowing node to start up..."
+        sleep 5
+        radicle-httpd --listen 0.0.0.0:8888 &
+        rad node status
+        rad self
+
+    - uses: pnpm/action-setup@v3
+      with:
+        version: 8
```
[the remark this answers is not in the document]
Comment (QZera) [reply .github/workflows/end-to-end-tests.yaml]: Added the `packageManager` key in package.json and updated relevant workflows to use pnpm v9

File: .github/workflows/end-to-end-tests.yaml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    - name: Install Node.js
+      uses: actions/setup-node@v4
```
[the remark this answers is not in the document]
Comment (QZera) [reply .github/workflows/end-to-end-tests.yaml]: Done!

File: .github/workflows/end-to-end-tests.yaml
Comment (maninak) [reply .github/workflows/end-to-end-tests.yaml:1]: Seems like GitHub was showing me outdated state of commits and comments for this PR, hence why my reply looks so off. :man_facepalming: 

TL;DR your initial observation on filenames misalignment was correct. As long as they are all aligned it's fine. I think it would make most sense if the filename is same as (using kebab-case) the workflow name.

File: package.json (hunk 2 above)
Comment (maninak) [reply package.json:53]: OK, great.

Resolving for now, but I think following up to make it possible to run e2e tests locally (and thus also out of GH actions!!) is worth coming back for a follow-up, given that it seems technically doable.

File: test/specs/onboarding.e2e.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
+import path from 'node:path'
+import { browser, expect } from '@wdio/globals'
+import { e2eTestDirPath } from 'test/constants/config'
+import type { ViewSection, Workbench } from 'wdio-vscode-service'
+import { $, cd } from 'zx'
+
+type VsCodeType = typeof import('vscode')
```
[the remark this answers is not in the document]
Comment (maninak) [reply test/specs/onboarding.e2e.ts:7]: The more I look at it the less important it gets. Leaving the call up to you.
Resolving.

File: test/specs/onboarding.e2e.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+  await radicleViewControl?.openView()
+}
+
+async function findFirstSectionWelcomeText(workbench: Workbench) {
+  const sidebarView = workbench.getSideBar().getContent()
+  await sidebarView.wait()
+  const firstSection = (await sidebarView.getSections())[0]
+  const welcomeContent = await firstSection?.findWelcomeContent()
+  const welcomeText = (await welcomeContent?.getTextSections()) ?? []
```
[the remark this answers is not in the document]
Comment (maninak) [reply test/specs/onboarding.e2e.ts:114]: I prefer `welcomeText3` which albeit not aesthetically pleasing, best communicates the data structure at a glance.

I prefer the `toMatchInlineSnapshot` approach.

The "RADICLE: GETTING STARTED" may be the `View.title` in case it helps. If not, it's fine.

**Review by maninak — CHANGES_REQUESTED** [review]
Follow-ups. Getting-there.

File: .github/workflows/end-to-end-tests.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (QZera) [reply .github/workflows/end-to-end-tests.yaml]: Removed the pnpm version lines, bumped the pnpm action version, from the workflows, also checked about caching we talked about and our setup is supposed to be already caching as expected.

File: test/specs/onboarding.e2e.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+describe('onboarding flow', () => {
+  let workbench: Workbench
+
+  before(async () => {
+    await setUpGitRepo()
+    workbench = await browser.getWorkbench()
+  })
+
+  describe('before the repo is initialized as a radicle repo', () => {
```
[the remark this answers is not in the document]
Comment (QZera) [reply test/specs/onboarding.e2e.ts]: oops, yeah I missed that. Added!

File: .vscode/settings.json [hunk 11]
```diff
@@ -46,4 +47,4 @@
     "source.fixAll.eslint": "explicit"
   },
   "git.alwaysSignOff": true,
-}
+}
```
[the remark this answers is not in the document]
Comment (QZera) [reply .vscode/settings.json]: Should be good now, as far as I can tell.

File: test/wdio.conf.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        userSettings: {
+          'extensions.autoCheckUpdates': false,
+          'extensions.autoUpdate': false,
+        },
+        vscodeArgs: {
+          'disable-extensions': true,
+          'disable-workspace-trust': true,
+        },
+      },
+      // Driver introduced in v9 has a bug that prevents tests from running
```
[the remark this answers is not in the document]
Comment (QZera) [reply test/wdio.conf.ts]: Going down the rabbit hole by searching with the error I'm getting leads me to [this issue](https://issues.chromium.org/issues/366886096) in Chromium. I don't know how much value it would have to add this link here, since the author of the issue seems to have encountered it while using Selenium.

**Review by maninak — CHANGES_REQUESTED** [review]
A few more things I came across.

Also, we're missing a changelog entry coming with this PR. Paste this above `## **v0.6.0** (November 9th, 2024)`:

```md
## _(WIP, yet unreleased version)_

### 🤖 CI

- **e2e:** set up new infrastructure and a related CI workflow capable of automated completely end-to-end testing. It uses a real Radicle node, which powers the latest extension build, which is running in an actual VS Code. Tests can assert behavior and state even as deep as extension webviews.

✅ Tests

- **onboarding:** cover various paths of the flow with e2e tests

-----

```

**Comment by QZera** [comment]
Regarding adding the "final new line" rule on workspace settings. I opted against doing it, since I saw that we don't want it for markdown files in `.editorconfig`.

**Review by QZera — COMMENTED** [review]
I think I got everything, please have a look!

File: package.json [hunk 13]
```diff
@@ -580,21 +580,38 @@
...
-    "typescript": "^5.3.3"
+    "tsx": "^4.19.2",
+    "typescript": "^5.3.3",
+    "wdio-vscode-service": "^6.1.2",
+    "webdriverio": "^9.2.11",
+    "zx": "^8.1.9"
+  },
+  "overrides": {
+    "wdio-vscode-service": {
+      "webdriverio": "^9.2.11"
```
[the remark this answers is not in the document]
Comment (maninak) [reply package.json:612]: I created this https://github.com/webdriverio-community/wdio-vscode-service/issues/140

Resolving with the work-around for now.
